config/config.go

```diff
@@ -267,6 +263,86 @@ func init() {
 	}
 }
 
+// SaveConfigurableConsensus saves the configurable protocols file to the provided data directory.
+func SaveConfigurableConsensus(dataDirectory string, params map[protocol.ConsensusVersion]ConsensusParams) error {
+	consensusProtocolPath := filepath.Join(dataDirectory, ConfigurableConsensusProtocolsFilename)
+
+	encodedConsensusParams, err := json.Marshal(params)
+	if err != nil {
+		return err
+	}
+	err = ioutil.WriteFile(consensusProtocolPath, encodedConsensusParams, 0644)
+	return err
+}
+
+// LoadConfigurableConsensusProtocols loads the configurable protocols from the data directroy
+func LoadConfigurableConsensusProtocols(dataDirectory string) error {
+	newConsensus, err := PreloadConfigurableConsensusProtocols(dataDirectory)
+	if err != nil {
+		return err
+	}
+	if newConsensus != nil {
+		Consensus = newConsensus
+	}
+	return nil
+}
+
+// PreloadConfigurableConsensusProtocols loads the configurable protocols from the data directroy
+// and merge it with a copy of the Consensus map. Then, it returns it to the caller.
+func PreloadConfigurableConsensusProtocols(dataDirectory string) (map[protocol.ConsensusVersion]ConsensusParams, error) {
+	consensusProtocolPath := filepath.Join(dataDirectory, ConfigurableConsensusProtocolsFilename)
+	file, err := os.Open(consensusProtocolPath)
+
+	if err != nil {
+		if os.IsNotExist(err) {
+			// this file is not required, only optional. if it's missing, no harm is done.
+			return nil, nil
+		}
+		return nil, err
+	}
+	defer file.Close()
+
+	configurableConsensus := make(map[protocol.ConsensusVersion]ConsensusParams)
+
+	decoder := json.NewDecoder(file)
+	err = decoder.Decode(&configurableConsensus)
+	if err != nil {
+		return nil, err
+	}
+
+	staticConsensus := make(map[protocol.ConsensusVersion]ConsensusParams)
+	for consensusVersion, consensusParams := range Consensus {
+		// recreate the ApprovedUpgrades map since we don't want to modify the original one.
+		if consensusParams.ApprovedUpgrades != nil {
+			newApprovedUpgrades := make(map[protocol.ConsensusVersion]uint64)
+			for ver, when := range consensusParams.ApprovedUpgrades {
+				newApprovedUpgrades[ver] = when
+			}
+			consensusParams.ApprovedUpgrades = newApprovedUpgrades
```

Comment:
consensusParams is a ref to the original/default entry in Consensus. what is the point of replacing its ApprovedUpgrades key by a new copy? Don't we simply want a new deep copy of Consensus map?

Reply:
consensusParams is not a ref; it's a copy. And since the ApprovedUpgrades is a map, we don't want to modify the original map but rather create a copy of it.
Do you have another way to generate a deep copy of the Consensus map ?